tsl/test/isolation/specs/cagg_watermark_concurrent_update.spec

```diff
@@ -0,0 +1,85 @@
+# This file and its contents are licensed under the Timescale License.
+# Please see the included NOTICE for copyright information and
+# LICENSE-TIMESCALE for a copy of the license.
+
+
+#
+# Test concurrent CAgg refreshes and invalidation threshold updates. This isolation test
+# checks that we don't skip CAgg updates when two sessions are trying to modify the
+# invalidation threshold at the same time.
+#
+setup
+{
+  CREATE TABLE temperature (
+    time timestamptz NOT NULL,
+    value float
+  );
+
+  SELECT create_hypertable('temperature', 'time');
+
+  INSERT INTO temperature
+    SELECT time, ceil(random() * 100)::int
+      FROM generate_series('2000-01-01 0:00:00+0'::timestamptz,
+                          '2000-01-01 23:59:59+0','1m') time;
+
+  CREATE MATERIALIZED VIEW cagg
+    WITH (timescaledb.continuous, timescaledb.materialized_only = false) AS
+    SELECT time_bucket('4 hour', time), avg(value)
+      FROM temperature
+      GROUP BY 1 ORDER BY 1
+    WITH NO DATA;
+}
+
+# Refresh CAGG in separate transaction
+setup
+{
+     CALL refresh_continuous_aggregate('cagg', NULL, NULL);
+}
+
+# Add new data to hypertable. This time in the year 2020 instead of 2000 as we
+# did for the setup of the CAgg.
+setup
+{
+  INSERT INTO temperature
+    SELECT time, ceil(random() * 100)::int
+      FROM generate_series('2020-01-01 0:00:00+0'::timestamptz,
+                          '2020-01-01 23:59:59+0','1m') time;
+}
+
+teardown {
+    DROP TABLE temperature CASCADE;
+}
+
+session "S1"
+step "s1_run_update" {
+   CALL refresh_continuous_aggregate('cagg', '2020-01-01 00:00:00', '2021-01-01 00:00:00');
+}
```

Comment:
Is it possible for the watermark to go back in time? Maybe if we truncate the cagg and refresh it to an earlier time point, or something like this. Would be good to also have a test for it.

Reply:
The test in the `cagg_query` doesn't check that we run the invalidations correctly, and a prepared statement that uses caggs is re-planned, right? I think it can be tested from normal SQL test using the `prepare`/`execute` SQL commands, probably with `plan_cache_mode = 'force_cached_plan'`.

Reply:
Correct. This part of the implementation is tested in `tsl/test/sql/cagg_watermark.sql.in`. We have a prepared statement in line 225 which is executed several times afterward.